File: llm_written/gemini-2.5-pro/gemm/c#/TestGemm.cs

using System;
using System.Diagnostics;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace GemmBenchmark
{
    public class TestGemm
    {
        private static Random rng = new Random(42);

        public static double[][] RandomMatrix(int rows, int cols)
        {
            var matrix = new double[rows][];
            for (int i = 0; i < rows; i++)
            {
                matrix[i] = new double[cols];
                for (int j = 0; j < cols; j++)
                {
                    matrix[i][j] = rng.NextDouble();
                }
            }
            return matrix;
        }

        public static string GetMatrixHash(double[][] matrix)
        {
            using (var sha256 = SHA256.Create())
            {
                var stringBuilder = new StringBuilder();
                for (int i = 0; i < matrix.Length; i++)
                {
                    for (int j = 0; j < matrix[i].Length; j++)
                    {
                        stringBuilder.Append(matrix[i][j].ToString("G17"));
                    }
                }
                byte[] bytes = Encoding.UTF8.GetBytes(stringBuilder.ToString());
                byte[] hash = sha256.ComputeHash(bytes);
                return BitConverter.ToString(hash).Replace("-", "").ToLowerInvariant();
            }
        }

        public static bool AreMatricesEqual(double[][] a, double[][] b, double tolerance = 1e-9)
        {
            if (a.Length != b.Length) return false;
            for (int i = 0; i < a.Length; i++)
            {
                if (a[i].Length != b[i].Length) return false;
                for (int j = 0; j < a[i].Length; j++)
                {
                    if (Math.Abs(a[i][j] - b[i][j]) > tolerance)
                    {
                        Console.WriteLine($"Mismatch at [{i}][{j}]: {a[i][j]} vs {b[i][j]}");
                        return false;
                    }
                }
            }
            return true;
        }

        public static void Main(string[] args)
        {
            var testCases = new[]
            {
                new { M = 1, K = 1, N = 1 },
                new { M = 16, K = 16, N = 16 },
                new { M = 63, K = 63, N = 63 },
                new { M = 64, K = 64, N = 64 },
                new { M = 128, K = 128, N = 128 },
                new { M = 256, K = 127, N = 255 },
                new { M = 512, K = 512, N = 512 }
            };

            var summary = new StringBuilder();
            bool allTestsPassed = true;

            foreach (var tc in testCases)
            {
                Console.WriteLine($"--- Testing {tc.M}x{tc.K} * {tc.K}x{tc.N} ---");
                summary.AppendLine($"--- Testing {tc.M}x{tc.K} * {tc.K}x{tc.N} ---");

                var A = RandomMatrix(tc.M, tc.K);
                var B = RandomMatrix(tc.K, tc.N);

                // Correctness Test
                var C_seq = Gemm.Run(A, B);
                var C_par = GemmParallel.Run(A, B);

                if (AreMatricesEqual(C_seq, C_par))
                {
                    Console.WriteLine("Correctness: PASS");
                    summary.AppendLine("Correctness: PASS");
                }
                else
                {
                    Console.WriteLine("Correctness: FAIL");
                    summary.AppendLine("Correctness: FAIL");
                    allTestsPassed = false;
                }

                // Determinism Test
                var hash1 = GetMatrixHash(GemmParallel.Run(A, B));
                var hash2 = GetMatrixHash(GemmParallel.Run(A, B));
                var hash3 = GetMatrixHash(GemmParallel.Run(A, B));

                if (hash1 == hash2 && hash2 == hash3)
                {
                    Console.WriteLine($"Determinism: PASS (hash: {hash1})");
                    summary.AppendLine($"Determinism: PASS (hash1: {hash1}, hash2: {hash2}, hash3: {hash3})");
                }
                else
                {
                    Console.WriteLine($"Determinism: FAIL (hashes: {hash1}, {hash2}, {hash3})");
                    summary.AppendLine($"Determinism: FAIL (hashes: {hash1}, {hash2}, {hash3})");
                    allTestsPassed = false;
                }
                summary.AppendLine();
            }
            
            // Performance Test
            int perf_M = 1024;
            int perf_K = 1024;
            int perf_N = 1024;
            Console.WriteLine($"--- Performance Test {perf_M}x{perf_K} * {perf_K}x{perf_N} ---");
            summary.AppendLine($"--- Performance Test {perf_M}x{perf_K} * {perf_K}x{perf_N} ---");
            var perfA = RandomMatrix(perf_M, perf_K);
            var perfB = RandomMatrix(perf_K, perf_N);

            var stopwatch = new Stopwatch();

            stopwatch.Restart();
            var perf_C_seq = Gemm.Run(perfA, perfB);
            stopwatch.Stop();
            var seq_time = stopwatch.Elapsed.TotalMilliseconds;
            Console.WriteLine($"Sequential time: {seq_time:F2} ms");
            summary.AppendLine($"Sequential time: {seq_time:F2} ms");

            stopwatch.Restart();
            var perf_C_par = GemmParallel.Run(perfA, perfB);
            stopwatch.Stop();
            var par_time = stopwatch.Elapsed.TotalMilliseconds;
            Console.WriteLine($"Parallel time:   {par_time:F2} ms");
            summary.AppendLine($"Parallel time:   {par_time:F2} ms");

            if (!AreMatricesEqual(perf_C_seq, perf_C_par))
            {
                 Console.WriteLine("Performance test correctness: FAIL");
                 summary.AppendLine("Performance test correctness: FAIL");
                 allTestsPassed = false;
            }

            double speedup = seq_time / par_time;
            Console.WriteLine($"Speedup: {speedup:F2}x");
            summary.AppendLine($"Speedup: {speedup:F2}x");
            
            var perfSummary = new StringBuilder();
            perfSummary.AppendLine($"M={perf_M}, K={perf_K}, N={perf_N}");
            perfSummary.AppendLine($"t_seq={seq_time:F2}ms");
            perfSummary.AppendLine($"t_par={par_time:F2}ms");
            perfSummary.AppendLine($"speedup={speedup:F2}x");
            perfSummary.AppendLine($"cores={Environment.ProcessorCount}");
            
            System.IO.File.WriteAllText("perf.txt", perfSummary.ToString());
            System.IO.File.WriteAllText("run_summary.txt", summary.ToString());

            if (!allTestsPassed)
            {
                Console.WriteLine("\nSome tests failed.");
                Environment.Exit(1);
            }
            else
            {
                Console.WriteLine("\nAll tests passed.");
            }
        }
    }
}

File: llm_written/gemini-2-5-pro/gemm/java/TestGemm.java
import java.util.Arrays;
import java.util.Random;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.math.BigInteger;

public class TestGemm {

    private static final int WARMUP_ITERATIONS = 5;
    private static final int BENCHMARK_ITERATIONS = 10;

    public static double[][] randomMatrix(int rows, int cols, Random rand) {
        double[][] matrix = new double[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = rand.nextDouble();
            }
        }
        return matrix;
    }

    public static String hashMatrix(double[][] matrix) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            for (double[] row : matrix) {
                for (double val : row) {
                    md.update(Double.toString(val).getBytes());
                }
            }
            byte[] digest = md.digest();
            return new BigInteger(1, digest).toString(16);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    public static boolean compareMatrices(double[][] a, double[][] b) {
        if (a.length != b.length || a[0].length != b[0].length) {
            return false;
        }
        for (int i = 0; i < a.length; i++) {
            for (int j = 0; j < a[0].length; j++) {
                if (Math.abs(a[i][j] - b[i][j]) > 1e-9) {
                    return false;
                }
            }
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println("Running GEMM correctness and determinism tests...");
        StringBuilder summary = new StringBuilder();

        // Test cases
        int[][] testShapes = {
            {1, 1, 1},       // Edge case: 1x1
            {10, 10, 10},    // Small
            {128, 128, 128}, // Medium
            {512, 512, 512}, // Large
            {100, 200, 300}  // Rectangular
        };

        boolean allTestsPassed = true;

        for (int[] shape : testShapes) {
            int m = shape[0], k = shape[1], n = shape[2];
            System.out.println("\nTesting shape: M=" + m + ", K=" + k + ", N=" + n);
            summary.append("\n--- Shape: M=").append(m).append(", K=").append(k).append(", N=").append(n).append(" ---\n");

            Random rand = new Random(12345); // Fixed seed for reproducibility
            double[][] A = randomMatrix(m, k, rand);
            double[][] B = randomMatrix(k, n, rand);

            // Run sequential version
            double[][] C_seq = Gemm.run(A, B);
            String hash_seq = hashMatrix(C_seq);
            summary.append("Sequential Hash: ").append(hash_seq).append("\n");

            // Run parallel version (correctness)
            double[][] C_par1 = GemmParallel.run(A, B);
            String hash_par1 = hashMatrix(C_par1);
            summary.append("Parallel Run 1 Hash: ").append(hash_par1).append("\n");

            boolean correct = compareMatrices(C_seq, C_par1);
            if (correct) {
                System.out.println("Correctness PASSED");
                summary.append("Correctness: PASSED\n");
            } else {
                System.out.println("Correctness FAILED");
                summary.append("Correctness: FAILED\n");
                allTestsPassed = false;
            }

            // Run parallel version again (determinism)
            double[][] C_par2 = GemmParallel.run(A, B);
            String hash_par2 = hashMatrix(C_par2);
            summary.append("Parallel Run 2 Hash: ").append(hash_par2).append("\n");

            boolean deterministic = hash_par1.equals(hash_par2);
            if (deterministic) {
                System.out.println("Determinism PASSED");
                summary.append("Determinism: PASSED\n");
            } else {
                System.out.println("Determinism FAILED");
                summary.append("Determinism: FAILED\n");
                allTestsPassed = false;
            }
        }

        System.out.println("\n--- Performance Tests ---");
        summary.append("\n--- Performance ---\n");
        StringBuilder perfSummary = new StringBuilder();
        perfSummary.append("shape,type,time_ms\n");


        int perf_m = 1024;
        int perf_k = 1024;
        int perf_n = 1024;

        Random rand = new Random(54321);
        double[][] A_perf = randomMatrix(perf_m, perf_k, rand);
        double[][] B_perf = randomMatrix(perf_k, perf_n, rand);

        // Warmup
        for (int i = 0; i < WARMUP_ITERATIONS; i++) {
            Gemm.run(A_perf, B_perf);
            GemmParallel.run(A_perf, B_perf);
        }

        // Sequential benchmark
        long totalSeqTime = 0;
        for (int i = 0; i < BENCHMARK_ITERATIONS; i++) {
            long startTime = System.nanoTime();
            Gemm.run(A_perf, B_perf);
            long endTime = System.nanoTime();
            totalSeqTime += (endTime - startTime);
        }
        double avgSeqTime = (totalSeqTime / 1e6) / BENCHMARK_ITERATIONS;
        System.out.printf("Sequential average time: %.2f ms\n", avgSeqTime);
        summary.append("Sequential average time: ").append(String.format("%.2f", avgSeqTime)).append(" ms\n");
        perfSummary.append(String.format("%dx%dx%d,sequential,%.2f\n", perf_m, perf_k, perf_n, avgSeqTime));


        // Parallel benchmark
        long totalParTime = 0;
        for (int i = 0; i < BENCHMARK_ITERATIONS; i++) {
            long startTime = System.nanoTime();
            GemmParallel.run(A_perf, B_perf);
            long endTime = System.nanoTime();
            totalParTime += (endTime - startTime);
        }
        double avgParTime = (totalParTime / 1e6) / BENCHMARK_ITERATIONS;
        System.out.printf("Parallel average time:   %.2f ms\n", avgParTime);
        summary.append("Parallel average time:   ").append(String.format("%.2f", avgParTime)).append(" ms\n");
        perfSummary.append(String.format("%dx%dx%d,parallel,%.2f\n", perf_m, perf_k, perf_n, avgParTime));

        double speedup = avgSeqTime / avgParTime;
        System.out.printf("Speedup: %.2fx\n", speedup);
        summary.append("Speedup: ").append(String.format("%.2f", speedup)).append("x\n");

        try {
            java.nio.file.Files.write(java.nio.file.Paths.get("run_summary.txt"), summary.toString().getBytes());
            java.nio.file.Files.write(java.nio.file.Paths.get("perf.txt"), perfSummary.toString().getBytes());
        } catch (java.io.IOException e) {
            System.err.println("Failed to write summary files.");
            e.printStackTrace();
        }

        if (!allTestsPassed) {
            System.exit(1);
        }
    }
}
